aiosmtpd/controller.py

```diff
@@ -185,9 +184,100 @@ def _stop(self):
         for task in _all_tasks(self.loop):
             task.cancel()
 
-    def stop(self):
-        assert self._thread is not None, "SMTP daemon not running"
+    def stop(self, no_assert=False):
+        assert no_assert or self._thread is not None, "SMTP daemon not running"
         self.loop.call_soon_threadsafe(self._stop)
-        self._thread.join()
-        self._thread = None
+        if self._thread is not None:
+            self._thread.join()
+            self._thread = None
```

Comment:
We now can bypass the assert test. Helpful for external exception handling because at that time, thread is possibly not started yet.